### ISSUE TITLE
Docker Breaks with Tailscale Running

### ISSUE THREAD

**Issue by charles-d-burton** [tailscale/tailscale#3877]
### What is the issue?

On a Linux machine running tailscale at the same time as docker breaks docker networking.  This happens with tailscale setup both as an exit node and without being an exit node.  I suspect it's something with double NAT but haven't had a chance to really dig into it.  Every request times out, ICMP doesn't work, and DNS resolution fails.

### Steps to reproduce

Install tailscale and docker on an Ubuntu 20.04.3 host using docker's default instructions.  Then do the following:
```bash
docker run --rm curlimages/curl:7.81.0 -L -v https://142.251.45.14 
 % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current                                 Dload  Upload   Total   Spent    Left  Speed  0     0    0     0    0     0      0      0 --:--:--  0:00:04 --:--:--     0* Could not resolve host: www.google.com* Closing connection 0curl: (6) Could not resolve host: www.google.com
```

```bash
sudo tcpdump -i tailscale0 -nn | grep 142.251.45.14
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on tailscale0, link-type RAW (Raw IP), capture size 262144 bytes
17:06:42.334357 IP 100.68.54.40.54398 > 142.251.45.14.443: Flags [S], seq 1485597718, win 64240, options [mss 1460,sackOK,TS val 1836179883 ecr 0,nop,wscale 7], length 0
17:06:42.377871 IP 142.251.45.14.443 > 100.68.54.40.54398: Flags [S.], seq 2574399181, ack 1485597719, win 65535, options [mss 1430,sackOK,TS val 2133358151 ecr 1836179883,nop,wscale 8], length 0
17:06:42.377931 IP 142.251.45.14.443 > 172.17.0.2.54398: Flags [S.], seq 2574399181, ack 1485597719, win 65535, options [mss 1430,sackOK,TS val 2133358151 ecr 1836179883,nop,wscale 8], length 0
17:06:42.686402 IP 142.251.45.14.443 > 100.68.54.40.54398: Flags [S.], seq 2574399181, ack 1485597719, win 65535, options [mss 1430,sackOK,TS val 2133358459 ecr 1836179883,nop,wscale 8], length 0
17:06:42.686457 IP 142.251.45.14.443 > 172.17.0.2.54398: Flags [S.], seq 2574399181, ack 1485597719, win 65535, options [mss 1430,sackOK,TS val 2133358459 ecr 1836179883,nop,wscale 8], length 0
17:06:43.365274 IP 100.68.54.40.54398 > 142.251.45.14.443: Flags [S], seq 1485597718, win 64240, options [mss 1460,sackOK,TS val 1836180914 ecr 0,nop,wscale 7], length 0
17:06:43.411817 IP 142.251.45.14.443 > 100.68.54.40.54398: Flags [S.], seq 2574399181, ack 1485597719, win 65535, options [mss 1430,sackOK,TS val 2133359185 ecr 1836179883,nop,wscale 8], length 0
17:06:43.411853 IP 142.251.45.14.443 > 172.17.0.2.54398: Flags [S.], seq 2574399181, ack 1485597719, win 65535, options [mss 1430,sackOK,TS val 2133359185 ecr 1836179883,nop,wscale 8], length 0
```

```bash
 docker run --rm curlimages/curl:7.81.0 -L -v https://142.251.45.14
*   Trying 142.251.45.14:443...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0* Connected to 142.251.45.14 (142.251.45.14) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
*  CAfile: /cacert.pem
*  CApath: none
} [5 bytes data]
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
} [512 bytes data]
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0* TLSv1.3 (IN), TLS handshake, Server hello (2):
{ [122 bytes data]
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
{ [15 bytes data]
* TLSv1.3 (IN), TLS handshake, Certificate (11):
{ [909 bytes data]
* TLSv1.3 (OUT), TLS alert, unknown CA (560):
} [2 bytes data]
* SSL certificate problem: self signed certificate
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
* Closing connection 0
curl: (60) SSL certificate problem: self signed certificate
More details here: https://curl.se/docs/sslcerts.html

curl failed to verify the legitimacy of the server and therefore could not
establish a secure connection to it. To learn more about this situation and
how to fix it, please visit the web page mentioned above.
```

### Are there any recent changes that introduced the issue?

None that I'm aware of, this is all pretty new.

### OS

Linux

### OS version

Ubuntu 20.04

### Tailscale version

1.20.2

### Bug report

BUG-5f2281a1312470cb6258ce2a27c2cc7e0d5293077c0128ce8fc2e709f1506059-20220204220621Z-98fcadb960cd208a

**Comment by charles-d-burton** [comment]
For color too, this is a complete showstopper for using tailscale.  Docker being unable to call out for builds means that it can't be used.  I'm working on a repro on another machine today as well

**Comment by DentonGentry** [comment]
We're going to need to look at the bugreport more, a quick look this weekend wasn't obvious what was going wrong.

There must be something else going on: we have a large number of sites running tailscale on ubuntu while using docker.

**Comment by charles-d-burton** [comment]
Yeah, it seems weird to me too.  In fact I don't remember having an issue with it before and I've been using tailscale for months now.  I can give you a dump of my route table too if that helps, I am seeing that tailscale is overriding the `172.17.0.0/16` table that docker0 is using but I've been working on other things today.

**Comment by charles-d-burton** [comment]
I was able to reproduce this behavior on another machine as well.  One that had a much more recent installation of both docker and tailscale as well as a much simpler networking config overall.

**Comment by DentonGentry** [comment]
If you haven't already, please install 1.20.4 on the exit node. The rest of the devices can stay on whatever 1.20.x they are running, but the exit node DNS forwarding has a significant improvement in 1.21.x and in 1.20.4.

**Comment by charles-d-burton** [comment]
Both systems are on 1.20.4 and the failure still occurs.  I don't think it's actually DNS though, all external networking breaks from within docker containers meaning that even using an IP address directly fails/times out.

**Comment by virenrshah** [comment]
Our Linux users are having the same issue ...both exitnode and linux tailscale are at 1.20.4

**Comment by mbugert** [comment]
Similar issue here, I couldn't manage to set up two 1.20.4 clients (Ubuntu 20.04 and Debian 11) with one as the exit node when tailscale was dockerized. For both containers, host networking and net_admin/sys_module permissions were set. Pinging the exit node's tailnet IP from the client was possible, but DNS queries and HTTP requests ran into timeouts.
Not using docker (on both sides) made it work.

**Comment by eirik-schwenke** [comment]
Experience similar problem here, for example - with tailscale up and exit-node enabled, the following:

`docker run --rm -ti python:3.9 getent hosts python.org` hangs, but with tailscale off,  or up, but with no exit-node works as expected:
```sh
docker run --rm -ti python:3.9 getent hosts python.org
138.197.63.241  python.org
```

Client node running:
```sh
tailscaled --version
1.22.0
  tailscale commit: b76d21c3af5a33d64beaddf8400762fcf843aa6e
  other commit: 4784a730477d34bdac862e5717bc6eefc1a803b5
  go version: go1.17.7-tsfd6b85b80f

lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 20.04.4 LTS
Release:	20.04
Codename:	focal

uname -a
Linux blackbody 5.13.0-30-generic #33~20.04.1-Ubuntu SMP Mon Feb 7 14:25:10 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

docker --version
Docker version 20.10.12, build e91ed57
```

Exit node:

```sh
tailscaled --version
1.22.0
  tailscale commit: b76d21c3af5a33d64beaddf8400762fcf843aa6e
  other commit: 4784a730477d34bdac862e5717bc6eefc1a803b5
  go version: go1.17.7-tsfd6b85b80f

lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 20.04.4 LTS
Release:	20.04
Codename:	focal

Linux tinytail 5.11.0-41-generic #45~20.04.1-Ubuntu SMP Wed Nov 10 10:20:10 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by danlsgiga** [comment]
Same issue here! I have an internal DNS server set to 192.168.1.3 via Admin console and `Override Local DNS` enabled. This works fine when I'm connected but *not* using an exit node. As soon as I enable to use an exit node, DNS breaks because it can't reach my internal 192.168.1.3 DNS server anymore.

**Comment by jfinas1** [comment]
Same problem for me (trying with a Postgres container). Both devices on version 1.24.2.
As a side note, I have the DNS set to my pihole, with local override enabled.
Connecting to the container does not work when the exit node is enabled and some interesting behavior is shown:
tailscaled log shows entries that suggest the traffic that should reach the docker network routed through tailscale IP, which I believe to be the issue. They get blocked by tailscale's ACL rules:
```
open-conn-track: flow TCP <tailscale-ip-of-my-machine>:46118 > 172.17.0.2:5432 rejected due to acl
open-conn-track: flow TCP <tailscale-ip-of-my-machine>:46118 > 172.17.0.2:5432 rejected due to acl
open-conn-track: flow TCP <tailscale-ip-of-my-machine>:46118 > 172.17.0.2:5432 rejected due to acl
```
Also, the DNS IP changes to 100.100.100, which I believe to be the IP of tailscale's magicdns (which I don't have enabled).
(`BUG-af18634a89abad1a2451d3c57330655554cd0704f3c1fbe0bbc6413f832b11aa-20220523151634Z-2b3028913535b56f` for reference)

As soon as I stop using the exit node `tailscale up --reset`, I can reach the Postgres container again, now weird "blocked by acl" -logs appear and the DNS is the actual pihole IP instead of 100.100.100.

**Comment by jfinas** [comment]
Same problem for me (trying with a Postgres container). Both devices on version 1.24.2.
As a side note, I have the DNS set to my pihole, with local override enabled.
Connecting to the container does not work when the exit node is enabled and some interesting behavior is shown:
tailscaled log shows entries that suggest the traffic that should reach the docker network routed through tailscale IP, which I believe to be the issue. They get blocked by tailscale's ACL rules:
```
open-conn-track: flow TCP <tailscale-ip-of-my-machine>:46118 > 172.17.0.2:5432 rejected due to acl
open-conn-track: flow TCP <tailscale-ip-of-my-machine>:46118 > 172.17.0.2:5432 rejected due to acl
open-conn-track: flow TCP <tailscale-ip-of-my-machine>:46118 > 172.17.0.2:5432 rejected due to acl
```
Also, the DNS IP changes to 100.100.100, which I believe to be the IP of tailscale's magicdns (which I don't have enabled).
(`BUG-af18634a89abad1a2451d3c57330655554cd0704f3c1fbe0bbc6413f832b11aa-20220523151634Z-2b3028913535b56f` for reference)

As soon as I stop using the exit node `tailscale up --reset`, I can reach the Postgres container again, now weird "blocked by acl" -logs appear and the DNS is the actual pihole IP instead of 100.100.100.

(this was reposted, since I accidentally did not use my private GitHub account)

**Comment by alex-shumilov** [comment]
Hey guys,
I have been fighting the same issue for a couple of days until I realised that Tailscale updates routing table for all routes except for anything directly bound to localhost or machine IP.
Long story short, `--exit-node-allow-lan-access` key fixed it for me.

**Comment by XBeg9** [comment]
same issue here, losing DNS on 1.30 even without `exit-node` usage, running tailscale in the docker (github actions https://github.com/tailscale/github-action)

**Comment by XBeg9** [comment]
How to reproduce:

```
docker run -it --device /dev/net/tun --cap-add=NET_ADMIN summerwind/actions-runner-dind:v2.296.1-ubuntu-20.04 /bin/bash
```

Then install tailscale and `accept-routes` after a minute you potentially will lose DNS and internet connection inside container. (you can do `accept-dns=false`, it doesn't help)

**Comment by XBeg9** [comment]
Team, any updates / suggestions for the fix for this issue?

**Comment by jamesmstone** [comment]
I wonder if you need to do something like this https://www.linuxserver.io/blog/routing-docker-host-and-container-traffic-through-wireguard

I am also suffering this issue, but haven't had time to really look into it

**Comment by XBeg9** [comment]
> 

looks like traffic in this case will go only through the wireguard (tailscale)

**Comment by heneri** [comment]
Setting the MTU size on the docker and tailscale* interface to the same value seems to fix the problem for me.

On my computer tailscale0 interface had mtu 1280 and docker0 interface mtu 1500.

`/etc/docker/daemon.json`

`
{
  "mtu": 1280
}
`

restart docker daemon.

**Comment by jamesmstone** [comment]
Hello

I was wondering if  anyone had any luck setting the `MTU` inside github actions?

at the moment I have:

```yml
    - name: Tailscale
      uses: tailscale/github-action@main
      with:
          authkey: ${{ secrets.TAILSCALE_AUTHKEY }}
          args: " --exit-node=<redacted ip>"
    - name: curl
      run: |-
        echo "outside curl:"
        curl:7.81.0 -L -vvvv https://tailscale.com 
        docker pull curlimages/curl:7.81.0
        temp=$(mktemp)
        jq '. + {"mtu": 1280 }'  < "/etc/docker/daemon.json" > "$temp"
        sudo mv "$temp" "/etc/docker/daemon.json"
        cat "/etc/docker/daemon.json"
        sudo  systemctl restart docker
        echo "restarted"
        docker run --rm curlimages/curl:7.81.0 -L -vvvv https://tailscale.com 
```

Which seems to set the MTU correctly, however it still fails to connect. Whereas outside docker the `curl` works fine.

Thanks!

**Comment by yeungon** [comment]
Have same problem and: 

`sudo tailscale up --exit-node=IP --exit-node-allow-lan-access=true` works for me.

**Comment by jamesmstone** [comment]
thanks, that is what I needed:

```yml
    - name: restart docker with smaller MTU
      run: |-
        temp=$(mktemp)
        jq '. + {"mtu": 1280 }'  < "/etc/docker/daemon.json" > "$temp"
        sudo mv "$temp" "/etc/docker/daemon.json"
        sudo  systemctl restart docker
        echo "restarted"
    - name: Tailscale
      uses: tailscale/github-action@main
      with:
          authkey: ${{ secrets.TAILSCALE_AUTHKEY }}
          args: " --exit-node=IP --exit-node-allow-lan-access=true"

```

I also found it helpful to restart docker before connecting to tailscale - although not sure if that is required.

**Comment by wpyoga** [comment]
On tailscale 1.38.4, HTTP connections and outgoing ICMP (ping) works, but DNS still doesn't work. (i.e. I can browse [1.1.1.1/help](http://1.1.1.1/help), but not google.com)

Using the flag `--exit-node-allow-lan-access=true` to work around this issue works only if my laptop's DNS server is set to a local LAN IP address (usually the home router).

A better workaround is outlined in [this article](https://techulus.xyz/fixing-docker-mtu-for-private-networks-and-vpns/). What I did was first create a new bridge network:

```
docker network create --opt com.docket.network.driver.mtu=1280 mtu1280
```

Then specify that network when creating the container:

```
docker run ... --network=mtu1280 ... tailscale/tailscale
```

**Comment by mhaddon** [comment]
> sudo tailscale up --exit-node=IP --exit-node-allow-lan-access=true

This works for me, but if i recreate the docker containers, i need to rerun this command.

**Comment by DentonGentry** [comment]
> This works for me, but if i recreate the docker containers, i need to rerun this command.

The container needs to have persistent storage for settings to be preserved across a restart of the container.

`docker run -d -v /opt/path/for/my/container/workload:/var/lib ...`

**Comment by Renari** [comment]
I ran into this today where I was unable to use docker compose to initiate a local mongodb connection. I found that disabling the exit node fixes the issue, nothing else I've tried seems to work. This was a deal breaker at work since all of our developers use docker.

**Comment by maltokyo** [comment]
this is a shame, nobody seems to have a definitive answer how to make tailscale exit node coexist with docker containers on the same host..

**Comment by pcrockett** [comment]
I have tried both setting the MTU and allowing LAN access, but even then I have to do `tailscale down && tailscale up` to get it working.

**Comment by mvnixon** [comment]
I have the same problem

**Comment by ei-grad** [comment]
Routing is broken. Response to 172.17.0.x (container IP address) goes back to `tailscale0`.

**Comment by VictorioBerra** [comment]
Im having this same issue and pulling my hair out. I locally cannot access any containers in docker with exposed ports on 127.0.0.1 or localhost. I did the daemon.json and verified MTU matches and sill does not work. `sudo tailscale down` fixes my issue thus confirming its Tailscale. Anyone find a solution?

My Mint 22.04 linux laptop is running Tailscale on the bare metal and I am NOT using it as an Exit node.

Final edit, it seems like at some point I told Tailscale to use an exit node, this command fixed my issues: `sudo tailscale set --exit-node=` still unsure if that is expected or desired.

**Comment by whitequark** [comment]
I'm able to get outbound Docker network access to work properly with an exit node enabled if I run:

```
tailscale up --exit-node=<name> --exit-node-allow-lan-access=true --stateful-filtering=false
```

**Comment by VictorioBerra** [comment]
@whitequark Yeah turns out my laptop was using an exit node and did not allow LAN access. The CLI status makes it hard to see that it was using the exit node. Killed the exit node setting and everything started working.
